crates/re_query_cache/src/cache.rs

```diff
@@ -0,0 +1,364 @@
+    pub archetype_name: ArchetypeName,
+}
+
+impl CacheKey {
+    #[inline]
+    pub fn new(
+        store_id: impl Into<StoreId>,
+        entity_path: impl Into<EntityPath>,
+        timeline: impl Into<Timeline>,
+        archetype_name: impl Into<ArchetypeName>,
+    ) -> Self {
+        Self {
+            store_id: store_id.into(),
+            entity_path: entity_path.into(),
+            timeline: timeline.into(),
+            archetype_name: archetype_name.into(),
+        }
+    }
+}
+
+// ---
+
+/// Caches the results of any query for an arbitrary range of time.
+///
+/// This caches all the steps involved in getting data ready for space views:
+/// - index search,
+/// - instance key joining,
+/// - deserialization.
+///
+/// We share the `CacheBucket` implementation between all types of queries to avoid duplication of
+/// logic, especially for things that require metaprogramming, to keep the macro madness to a
+/// minimum.
+/// In the case of `LatestAt` queries, a `CacheBucket` will always contain a single timestamp worth
+/// of data.
+#[derive(Default)]
+pub struct CacheBucket {
+    /// The timestamps and [`RowId`]s of all cached rows.
+    ///
+    /// Reminder: within a single timestamp, rows are sorted according to their [`RowId`]s.
+    pub(crate) pov_times: VecDeque<(TimeInt, RowId)>,
+
+    /// The [`InstanceKey`]s of the point-of-view components.
+    pub(crate) pov_instance_keys: FlatVecDeque<InstanceKey>,
+
+    /// The resulting component data, pre-deserialized, pre-joined.
+    //
+    // TODO(#4733): Don't denormalize auto-generated instance keys.
+    // TODO(#4734): Don't denormalize splatted values.
+    pub(crate) components: BTreeMap<ComponentName, Box<dyn ErasedFlatVecDeque + Send + Sync>>,
+    //
+    // TODO(cmc): secondary cache
+    // TODO(#4730): size stats: this requires codegen'ing SizeBytes for all components!
+}
+
+impl CacheBucket {
+    /// Iterate over the timestamps of the point-of-view components.
+    #[inline]
+    pub fn iter_pov_times(&self) -> impl Iterator<Item = &(TimeInt, RowId)> {
+        self.pov_times.iter()
+    }
+
+    /// Iterate over the [`InstanceKey`] batches of the point-of-view components.
+    #[inline]
+    pub fn iter_pov_instance_keys(&self) -> impl Iterator<Item = &[InstanceKey]> {
+        self.pov_instance_keys.iter()
+    }
+
+    /// Iterate over the batches of the specified non-optional component.
+    #[inline]
+    pub fn iter_component<C: Component + Send + Sync + 'static>(
+        &self,
+    ) -> re_query::Result<impl Iterator<Item = &[C]>> {
+        let data = self
+            .components
+            .get(&C::name())
+            .map(|data| data.as_any().downcast_ref::<FlatVecDeque<C>>())
+            .ok_or_else(|| re_query::QueryError::RequiredComponentNotFound(C::name()))?;
+
+        let Some(data) = data else {
+            return Ok(Either::Left(std::iter::empty()));
+        };
+
+        Ok(Either::Right(data.iter()))
```

Comment:
nit/style: I find this easier to read

```suggestion
        if let Some(data) = data {
            Ok(Either::Right(data.iter()))
        } else {
            Ok(Either::Left(std::iter::empty()))
        };
```